[BREAKING] tx.origin, msg.sender and address(literal) are non-payable addresses (#10496)
ethereum/solidity: docs/types/value-types.rst

Implicit conversions from ``address payable`` to ``address`` are allowed, whereas conversions from ``address`` to ``address payable``
must be explicit via ``payable(<address>)``.

Explicit conversions to and from ``address`` are allowed for integers, integer literals, ``bytes20`` and contract types with the following
caveat:
The result of a conversion of the form ``address(x)``

> has the type ``address payable``, if ``x`` is of integer or fixed bytes type,

Reviewer: Still needs to be changed.

> or a contract with a receive or payable fallback function.

Reviewer 2: The documentation change looks good otherwise (apart from the one other phrase I commented about), but this here made me stop and ask, if it still being ``address payable`` for integers and fixed bytes types is actually what we want...

Reviewer: No we don't.

If ``x`` is a contract without a receive or payable fallback function,
then ``address(x)`` will be of type ``address``.
Similarly, if ``x`` is a literal, then ``address(x)`` will also be of type ``address``.
In external function signatures ``address`` is used for both the ``address`` and the ``address payable`` type.

Only expressions of type ``address`` can be converted to type ``address payable`` via ``payable(<address>)``.

.. note::
    If you store a variable of ``address`` type and plan to send Ether to it, then
    store it as ``address payable`` to make this requirement visible. Also try to pull this

>     type as far out as possible or do the conversion as early as possible.

Reviewer 2: Not sure pulling the type far out is that easy to understand... do you basically mean "Prefer the type ``address payable``, whenever you can be sure that an ``address`` is in fact ``payable``" or something like that?

>     If you store a variable of ``address`` type and plan to send Ether to it, then
>     store it as ``address payable`` to make this requirement visible. Also try to pull this
>     type as far out as possible or do the conversion as early as possible.

Author: ```suggestion
    If you need a variable of type ``address`` and plan to send Ether to it, then
    declare its type as ``address payable`` to make this requirement visible. Also, 
    try to make this distinction or conversion as early as possible.
```

Author: ?

Reviewer 2: I like that!

    and ``address payable`` and just use ``address`` everywhere. For example,

>     if you are using the :ref:`withdrawal pattern<withdrawal_pattern>`, you can (and should) store the

Reviewer: I think we just contradicted this paragraph with the example above, right?

The thing I learnt from the example is that you should do the conversion to address payable as early as possible, so that it is more likely that a caller / user of a function sees that the address might receive ether in the future.

Author: > I think we just contradicted this paragraph with the example above, right?

Yeah, I agree. I'm not sure why we don't want to have `owner` in the withdrawal example to be a payable address. IMO, we should replace `payable(msg.sender).transfer...` by `owner.transfer...` whenever possible.
